driver/adfs_proxy.cc

```diff
@@ -28,31 +28,245 @@
+  this->http_client =
+      std::make_shared<SAML_HTTP_CLIENT>("https://" + host, connect_timeout, socket_timeout, enable_ssl);
+}
+
+std::string ADFS_SAML_UTIL::get_saml_assertion(DataSource* ds) {
+  nlohmann::json res;
+  try {
+    res = this->http_client->get(std::string(SIGN_IN_PAGE_URL));
+  } catch (SAML_HTTP_EXCEPTION& e) {
+    const std::string error =
+        "Failed to get sign-in page from ADFS: " + e.error_message() + ". Please verify your IDP endpoint.";
+    throw SAML_HTTP_EXCEPTION(error);
+  }
+
+  const auto body = std::string(res);
+  std::smatch m;
+  if (!std::regex_search(body, m, ADFS_REGEX::FORM_ACTION_PATTERN)) {
+    return std::string();
   }
-  else {
-      this->auth_util = std::make_shared<AUTH_UTIL>();
+  std::string form_action = escape_html_entity(m.str(1));
+  const std::string params = get_parameters_from_html(ds, body);
+  const std::string content = get_form_action_body(form_action, params);
+  printf("hjello %s\n", content.c_str());
+  if (std::regex_search(content, m, ADFS_REGEX::SAML_RESPONSE_PATTERN)) {
+    return m.str(1);
   }
+  return std::string();
+}
+
+std::string ADFS_SAML_UTIL::escape_html_entity(const std::string& html) {
+  std::string retval("");
+  int i = 0;
+  int length = html.length();
+  while (i < length) {
+    char c = html[i];
+    if (c != '&') {
+      retval.append(1, c);
+      i++;
+      continue;
+    }
+
+    if (html.substr(i, 4) == "&lt;") {
+      retval.append(1, '<');
+      i += 4;
+    } else if (html.substr(i, 4) == "&gt;") {
+      retval.append(1, '>');
+      i += 4;
+    } else if (html.substr(i, 5) == "&amp;") {
+      retval.append(1, '&');
+      i += 5;
+    } else if (html.substr(i, 6) == "&apos;") {
+      retval.append(1, '\'');
+      i += 6;
+    } else if (html.substr(i, 6) == "&quot;") {
+      retval.append(1, '"');
+      i += 6;
+    } else {
+      retval.append(1, c);
+      ++i;
+    }
+  }
+  return retval;
+}
+
+std::vector<std::string> ADFS_SAML_UTIL::get_input_tags_from_html(const std::string& body) {
+  std::unordered_set<std::string> hashSet;
+  std::vector<std::string> retval;
+
+  std::smatch matches;
+  std::regex pattern(ADFS_REGEX::INPUT_TAG_PATTERN);
+  std::string source = body;
+  while (std::regex_search(source, matches, pattern)) {
+    std::string tag = matches.str(0);
+    std::string tagName = get_value_by_key(tag, std::string("name"));
+    std::transform(tagName.begin(), tagName.end(), tagName.begin(), [](unsigned char c) { return std::tolower(c); });
+    if (!tagName.empty() && hashSet.find(tagName) == hashSet.end()) {
+      hashSet.insert(tagName);
+      retval.push_back(tag);
+    }
+
+    source = matches.suffix().str();
+  }
+
+  return retval;
+}
+
+std::string ADFS_SAML_UTIL::get_value_by_key(const std::string& input, const std::string& key) {
+  std::string pattern("(");
+  pattern += key;
+  pattern += ")\\s*=\\s*\"(.*?)\"";
```

Comment:
Minor: `std::regex pattern` on line 129 above is using a `const` for the pattern. Does it make sense to define a `const` for this pattern as well?